src/r_main.c

```diff
@@ -638,14 +639,38 @@ void R_SetupFrame (player_t *player)
       // Don't interpolate during a paused state
       leveltime > oldleveltime)
   {
+    const boolean use_localview = (
+      // Don't use localview if the player is spying.
+      player == &players[consoleplayer] &&
+      // Don't use localview if the player is dead.
+      player->health > 0 &&
+      // Don't use localview if the player just teleported.
+      !player->mo->reactiontime &&
+      // Don't use localview if a demo is playing.
+      !demoplayback &&
+      // Don't use localview during a netgame (single-player and solo-net only).
+      (!netgame || solonet)
+    );
+
     // Interpolate player camera from their old position to their current one.
     viewx = player->mo->oldx + FixedMul(player->mo->x - player->mo->oldx, fractionaltic);
     viewy = player->mo->oldy + FixedMul(player->mo->y - player->mo->oldy, fractionaltic);
     viewz = player->oldviewz + FixedMul(player->viewz - player->oldviewz, fractionaltic);
-    viewangle = R_InterpolateAngle(player->mo->oldangle, player->mo->angle, fractionaltic) + viewangleoffset;
+
+    // Use localview unless the player or game is in an invalid state or if
+    // mouse input was interrupted, in which case fall back to interpolation.
+    if (localview.useangle && use_localview)
+      viewangle = player->mo->angle - ((short)localview.angle << FRACBITS) + viewangleoffset;
+    else
+      viewangle = R_InterpolateAngle(player->mo->oldangle, player->mo->angle, fractionaltic) + viewangleoffset;
+
+    if (localview.usepitch && use_localview && !player->centering && player->lookdir)
+      pitch = (player->lookdir + localview.pitch) / MLOOKUNIT;
+    else
+      pitch = (player->oldlookdir + (player->lookdir - player->oldlookdir) * FIXED2DOUBLE(fractionaltic)) / MLOOKUNIT;
```

Comment:
Here's a visualization. The white arrow is `viewangle` which is what the player sees on the screen. The green arrow is `player->mo->angle` which is the actual angle used during a game simulation update. With fast mouse polling, the game sim lags behind the screen and mouse input. With regular interpolation, the screen lags behind the game sim and mouse input. In either case, everything syncs up every 1/35 s as expected. In the video I slowed down the game speed and exaggerated the effect to make it easier to see.

https://github.com/fabiangreffrath/woof/assets/56656010/03ea694a-2448-4764-82b7-9edaf5c9bdbf

<details><summary>Code for arrows:</summary>

```diff
diff --git a/src/am_map.c b/src/am_map.c
index c84131d0..5f848dc5 100644
--- a/src/am_map.c
+++ b/src/am_map.c
@@ -1950,6 +1950,17 @@ static void AM_drawPlayers(void)
         pt.y
       ); 
     else
+    {
+      AM_drawLineCharacter
+      (
+        player_arrow,
+        NUMPLYRLINES,
+        0,
+        plr->mo->angle,
+        mapcolor_plyr[0],
+        pt.x,
+        pt.y);
+
       AM_drawLineCharacter
       (
         player_arrow,
@@ -1959,6 +1970,7 @@ static void AM_drawPlayers(void)
         mapcolor_sngl,      //jff color
         pt.x,
         pt.y);        
+    }
     return;
   }
```

</details>

This is a pretty major change in how the mouse input will feel. Before merging this, should there be a setting to toggle it? Maybe config only? Note that it doesn't work with multiplayer or short tics so the slower feeling mouse will still be present there.

Reply:
Thanks for the demonstration! The mouse will feel super responsive and snappy. Honestly, I don't think we'll need a "I want my mouse to lag behind again" config switch.